lineapy/plugins/taskgen.py

```diff
@@ -0,0 +1,221 @@
+"""
+This taskgen files contains helper functions to create tasks.
+"""
+from typing import Dict
+
+from lineapy.graph_reader.artifact_collection import ArtifactCollection
+from lineapy.graph_reader.node_collection import ArtifactNodeCollection
+from lineapy.plugins.session_writers import BaseSessionWriter
+from lineapy.plugins.task import DagTaskBreakdown, TaskDefinition
+from lineapy.plugins.utils import load_plugin_template
+
+
+def get_task_definitions(
+    artifact_collection: ArtifactCollection,
+    pipeline_name: str,
+    task_breakdown: DagTaskBreakdown,
+) -> Dict[str, TaskDefinition]:
+    """
+    get_task_definitions returns a dictionary of TaskDefinitions
+
+    This function breaks down the artifact_collection into tasks based on the
+    task_breakdown parameter. This will give the main bulk of tasks that should
+    be included in a pipeline dag file.
+
+    Returns a `task_definitions` dictionary, which maps a key corresponding to the task name to
+    Linea's TaskDefinition object.
+    Specific framework implementations of PipelineWriters should serialize the TaskDefinition
+    objects to match the format for pipeline arguments that is expected by that framework.
+    """
+    if task_breakdown == DagTaskBreakdown.TaskAllSessions:
+        return get_allsessions_task_definition(pipeline_name)
+    elif task_breakdown == DagTaskBreakdown.TaskPerSession:
+        return get_session_task_definition(artifact_collection, pipeline_name)
+    elif task_breakdown == DagTaskBreakdown.TaskPerArtifact:
+        return get_artifact_task_definitions(
+            artifact_collection, pipeline_name
+        )
+    else:
+        raise ValueError(
+            f"Task breakdown granularity {task_breakdown} is not currently supported."
+        )
+
+
+def get_artifact_task_definitions(
+    artifact_collection: ArtifactCollection, pipeline_name: str
+) -> Dict[str, TaskDefinition]:
+    """
+    get_artifact_task_definitions returns a task definition for each artifact the pipeline produces.
+    This may include tasks that produce common variables that were not initially defined as artifacts.
+    """
+    task_definitions: Dict[str, TaskDefinition] = dict()
+    unused_input_parameters = set(artifact_collection.input_parameters)
+
+    for session_artifacts in artifact_collection.sort_session_artifacts():
```

Comment:
We should allow for user-provided `dependencies: TaskGraphEdge` in `artifact_collection.sort_session_artifacts()` to account for multi-session pipeline cases.

For this, `dependencies` parameter should be added to:

```python
def get_task_definitions(
    artifact_collection: ArtifactCollection,
    pipeline_name: str,
    task_breakdown: DagTaskBreakdown,
) -> Dict[str, TaskDefinition]:
```

Reply:
Sure, will include this in the new ticket. I agree with this too.